Added baseFile optional flag

harryjoy opened the pull request:
As a solution to issue #1, 
Added baseFile optional flag and if that's true then don't rename that folder and just copy as it is to destination.

harryjoy committed:
Added baseFile optional flag and if that's true then don't rename that folder and just copy as it is to destination.

harryjoy commented:
Added an option called `isOptional` when baseFile is not found in folder, it checks for this flag and if its false then it throws error otherwise it continues to process with old name.

In case of isOptional = true and no baseFile, renameTo function will receive baseFileData as undefined or null and second argument has been added that represent old folder name (current name while processing) that can be used by implementor to process in such cases.

Example is:

```
      subdirRename({
            baseFile : "./module.json",
            isOptional: true,
            renameTo : function (baseFileData, oldSubDirName) {
                if (baseFileData) {
                    var moduleJson = JSON.parse(baseFileData);
                    return moduleJson.name;
                } else {
                    return "";
                }
            }
        })
```

if implementor returns "" (blank) or null or undefined as result of renameTo then old name is used and it will copied to destination as it is.

harryjoy commented:
It also solved problem when there is files and folders both in place where you want to rename. Currently it copies file as it is and rename all folders as configured; if `isOptional` flag is true with `baseFile`.

davelosert commented:
Hi·

sorry my late response - I was on a business-travel the whole last week and didn't have time.
Thank you very much for the PR. Im gonna look into this tonight or tomorrow.

Next steps then would be to merge and release a new version of the plugin.

davelosert commented:
@harryjoy : Code seems good. All that is missing is a test and the documentation. Are you up for doing this also or do you want me to do it?

A test could look like this:

```
it('should not throw if fs throws but isOptional option is true', function () {
        var fsStub = {
            readFile: sinon.stub().yields(new Error('File not found!'))
        };

        stream = proxyquire('../src/stream', {
            './mapFileFinder.js': mapFileFinderStub,
            'fs': fsStub
        });
        options.isOptional = true;
        myStream = stream(options);
        var funcWrapper = function () {
            myStream.write(testFile);
        };
        expect(funcWrapper).to.not.throw('gulp-subdir-rename - Error while reading baseFile: File not found!');
    })
```

harryjoy committed:
Adding test cases for isOptional flag

harryjoy commented:
Sure. I missed looking into test folder. I will do that and update this PR.

harryjoy committed:
Updated readme to add details about isOptional flag.

harryjoy commented:
Added test cases for `isOptional` flag and updated `readme.md` with details (I guess that what you meant by documents as I don't see any Wiki pages for repo)

davelosert commented on readme.md:
```diff
@@ -74,8 +74,9 @@ where the glob starts, so for example for a  `gulp.src('/base/path/**')` it woul
-It's parameter are the contents of the specified `baseFile` (as a string).
+* **renameTo** - *function(baseFileContent, oldSubDirName)* : The functions return value will be used as the new subdirectories name.
+It's first parameter are the contents of the specified `baseFile` (as a string) and second parameter is `old name of directory` that will be changed. If `""`or `undefined` or `null` is returned from the function then no rename is done and folder is copied as it is.
+* **isOptional** - *boolean* : When `baseFile` is not found in folder, processing checks for `isOptional` flag and if its `false` then it throws error otherwise it continues to process with old name. In case of `isOptional = true` and no `baseFile` present in folder, `renameTo` function will receive `baseFileContent` as `undefined` or `null`, in such case implementor can use second argument, that ...
```
This is a little bit long imho. Maybe we could change it to something like:
- **isOptional** - _boolean (default: false)_ : set this to true if you have folders which do not contain the baseFile. Per default, those folders would cause an error. If set to true, those folders will then just be copied as they are with their original dir-name.

harryjoy replied:
Yea I agree it's quite big but only reason I thought it would makes senses
because repo does not have wiki and this is only place to describe. But
yeah, if you feel like it's big; please feel free to cut it and merge.

Thanks

On Sunday 24 July 2016, David Losert notifications@github.com wrote:

> In readme.md
> https://github.com/davelosert/gulp-subdir-rename/pull/2#discussion_r71986202
> :
> 
> > @@ -74,8 +74,9 @@ where the glob starts, so for example for a  `gulp.src('/base/path/**')` it woul
> > - **baseFile** - _string_ : Path to the file you wanna pull the subdirectories name from. The Path must be relative
> >   to the subdirectory that's to be renamed (so the first directory after the base-path). It's content will be given as a
> >   paramter to `renameTo()`.
> >   -\* **renameTo** - _function(baseFileContent)_ : The functions return value will be used as the new subdirectories name.
> >   -It's parameter are the contents of the specified `baseFile` (as a string).
> >   +\* **renameTo** - _function(baseFileContent, oldSubDirName)_ : The functions return value will be used as the new subdirectories name.
> >   +It's first parameter are the contents of the specified `baseFile` (as a string) and second parameter is `old name of directory` that will be changed. If `""`or `undefined` or `null` is returned from the function then no rename is done and folder is copied as it is.
> >   +\* **isOptional** - _boolean_ : When `baseFile` is not found in folder, processing checks for `isOptional` flag and if its `false` then it throws error otherwise it continues to process with old name. In case of `isOptional = true` and no `baseFile` present in folder, `renameTo` function will receive `baseFileContent` as `undefined` or `null`, in such case implementor can use second argument, that is old name and process on it to generate and return new name to be used for the folder. When `isOpional = true` and there are files in base folder then those files will be copied as it is. Default value is `false`.
> 
> This is a little bit long imho. Maybe we could change it to something like:
> - _isOptional_ - _boolean (default: false)_ : set this to true if you
>   have folders which do not contain the baseFile. Per default, those folders
>   would cause an error. If set to true, those folders will then just be
>   copied as they are with their original dir-name.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/davelosert/gulp-subdir-rename/pull/2/files/51bc85aa95c4c9c7bf78eb90c3735040741104e7#r71986202,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAnkeL8CFJpygW5c7rDlGVMsugJx-QgYks5qYz9DgaJpZM4JJRfH
> .

davelosert commented:
@harryjoy : Awesome work so far, thank you very much.
One more minor thing with the documentation - then this PR is ready to be merged! :+1:
